Adding metrics to NetworkClient

nsivabalan opened the pull request:
Adding metrics to NetworkClient. 

SLA: 30 mins
Reviewers: Priyesh, Ming

Some notes to assist in reviewing:
Stage1: 10 mins
Renamed the class NetworkRequestMetircs to NetworkSendMetrics.
ClientNetworkRequestMetrics, ServerNetworkResponseMetrics inherits from NetworkSendMetrics

Stage2: 15 mins
NetworkClient and NetworkMetrics
This has the major change in this patch. Introduced new metrics to track metrics for NetworkClient. Also have rearranged existing metrics in NetworkMetrics a bit.
Misc classes: NetworkClientFactory, NetworkClientTest, MockNetworkClientFactory,

vgkholla commented on ambry-api/src/main/java/com.github.ambry/network/ClientNetworkRequestMetrics.java:
```diff
@@ -0,0 +1,32 @@
+ * Tracks a set of metrics for a network request by a client
+ */
+public class ClientNetworkRequestMetrics implements NetworkRequestSend {
+  private MetricsHistogram requestSendTime;
```
we usually use `Histogram` from codahale directly. Why use this? Its implementation is eventually using a `Histogram` anyway.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/network/NetworkRequestSend.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * An interface to update send time for a network request
+ */
+public interface NetworkRequestSend {
```
not sure why you need this interface.

xiahome replied:
Why you have this interface, and why you name it like this? Will it be only used to collect metrics for `NetworkSend`?

xiahome replied:
ah, same question as @vgkholla has.

nsivabalan replied:
If you take a look at how NetworkSend is created at the server side, (line 476 in SocketServer)[https://github.com/linkedin/ambry/blob/master/ambry-network/src/main/java/com.github.ambry.network/SocketServer.java], we don't have an handle to every metric in particular. Its abstracted in a class which exposes method called updateSendTime(value). It might take some effort to fix this. But I don't want to touch the server code for now (since its been running smoothly until now). So, modeled the client also to implement the same api. May be we can sync up offline if you have more doubts

vgkholla replied:
Still doesn't explain why we need interfaces for this. If you
1. Change the constructor of `NetworkRequestMetrics` to take in `Histogram` instead of `MetricsHistogram`.
2. Change the function names inside `NetworkRequestMetrics` to `updateQueueTime()` and `updateSendTime()`.

and then used the concrete class `NetworkRequestMetrics` and instantiated it with the right histograms, I don't see why it cannot be used both on the server and client side. Are there any problems with that?

vgkholla replied:
You don't even need to do (1) - but I would like that anyway because I don't see the value of `MetricsHistogram`.

vgkholla commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSend.java:
```diff
@@ -36,9 +36,9 @@
+  private final NetworkRequestSend metrics;
 
-  public NetworkSend(String connectionId, Send payload, NetworkRequestMetrics metrics, Time time) {
+  public NetworkSend(String connectionId, Send payload, NetworkRequestSend metrics, Time time) {
```
why not just send a metrics object like the others that we define and simply update the actual histogram required?

vgkholla replied:
What is the drawback with `NetworkRequestMetrics`?

vgkholla commented on ambry-api/src/main/java/com.github.ambry/network/ServerNetworkRequestMetrics.java:
```diff
@@ -17,17 +17,17 @@
+ * Tracks a set of metrics for a network request by a server
  */
-public class NetworkRequestMetrics {
+public class ServerNetworkRequestMetrics implements NetworkRequestSend {
```
why the rename?

vgkholla commented on ambry-network/src/main/java/com.github.ambry.network/HistogramMeasurement.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Class to assist in updating the {@link Histogram} with a value
+ */
+public class HistogramMeasurement implements MetricsHistogram {
```
don't see why this is required.

vgkholla commented:
I might be wrong, but I feel like this patch could be *much* simpler. I don't see why we need all the extra interfaces and implementations. Have you guys discussed something offline that I don't have context about?

In fact, before we tackle this, I think we should first change `NetworkRequestMetrics` to start taking in `Histogram` (from codahale). I'm unable to see or explain what value `MetricsHistogram` adds.

pnarayanan commented:
The existing code uses `MetricsHistogram`. I though this was required to save the reference of the metric passed in. Siva and I had noticed it and finding out the value of `MetricsHistogram` in `NetworkRequestMetrics` was part of this task.

vgkholla commented:
Not sure what you mean but if we replaced all the `MetricsHistogram` in the constructor of `NetworkRequestMetrics` with `Histogram`, it should still work.

pnarayanan commented:
It appears so, yes.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/ClientNetworkRequestMetrics.java:
```diff
@@ -0,0 +1,32 @@
+
+
+/**
+ * Tracks a set of metrics for a network request by a client
```
I suggest this doc to be improved. I guess by "network request" you mean `NetworkSend`. It is also not clear what does client mean here.

xiahome commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -30,23 +31,30 @@
+  public final Meter connectionCheckOutAttemptsCount;
+  // number of times connection checkout have been tried before timing out
+  public final Meter connectionCheckOutFailureAttemptsCount;
+  public final Meter connectionsDisconnectedCount;
```
Name is confusing, maybe `connectionDisconnectingRate`?

nsivabalan replied:
Already we have metrics in similar names. So going with the same. Do check NettyMetrics for connectionsDisconnectedCount

xiahome commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -30,23 +31,30 @@
+  public final Counter sslRenegotiationCount;
+  // NetworkClient metrics
+  // number of times connection checkout have been tried before succeeding
+  public final Meter connectionCheckOutAttemptsCount;
```
Name is confusing, maybe `connectionAttemptedCheckoutRate`?

xiahome commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -30,23 +31,30 @@
+  // number of times connection checkout have been tried before succeeding
+  public final Meter connectionCheckOutAttemptsCount;
+  // number of times connection checkout have been tried before timing out
+  public final Meter connectionCheckOutFailureAttemptsCount;
```
Name is confusing, maybe `connectionCheckoutFailureRate`?

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSend.java:
```diff
@@ -59,7 +51,7 @@ public Send getPayload() {
   public void onSendComplete() {
     if (metrics != null) {
-      metrics.updateResponseSendTime(SystemTime.getInstance().milliseconds() - sendStartTimeInMs);
+      metrics.updateSendTime(time.milliseconds() - sendStartTimeInMs);
```
Why is this even `updateSendTime()`? Since `NetworkSendMetrics` is a black box to `NetworkSend`, this should probably be `metrics.onSendComplete()`. It even looks like the `NetworkRequestMetrics` object should contain the start time as well. Depending on the use case, it should either be provided the start time, or `NetworkSend` should call a method `metrics.onSendStart()` (looking at the code it seems like the former is good enough).

nsivabalan replied:
Let me try to explain my take here. NetworkSendMetrics has two apis, updateQueueTime(long) and updateSendTime(long). Even I was thinking that we should initialize the starting time when a new NetworkSendMetrics is constructed and when onDequeu(), we should subtract the time and update the queue time. similarly onSendComplete(), should subtract from dequeue time and update the send time. But looks like queueing could have been done way before NetworkSendMetrics are constructed and hence updateQueueTime() takes in a value to be updated. So, I thought I would keep the updateSendTime() to be same lines w/o changing anything.

nsivabalan replied:
But having said that, I am open to change the updateSendTime(long) to onSendComplete(). But then, I need to have the "Time" instance, to get the millisecs and might need some changes to the NetworkSendMetrics.

pnarayanan replied:
what is `updateQueueTime()`? Does the `NetworkSend` care about it?

nsivabalan replied:
Yes, its the time from which the NetworkSend was queued up until it was dequeued to be sent via selector. From client's perspective, this is time from which the networkClient receives a new send from the caller of sendAndPoll(). For Server, it is the time at which the "Send" was sent to RequestResponseChannel(i.e. SocketRequestResponseChannel).

pnarayanan replied:
`NetworkSend` does not invoke `updateQueueTime()`.

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,65 @@
+  public void updateSendTime(long value) {
+    requestOrResponseSendTime.update(value);
+    if (requestOrResponseSendTimeBySize != null) {
+      requestOrResponseSendTimeBySize.update(value);
```
I don't quite see the difference between `requestOrResponseSendTime` and `requestOrResponseSendTimeBySize`. The same goes for `requestOrResponseTotalTime` and `requestOrResponseTotalTimeBySize`. They seem to be updated exactly the same way. Do you know?

nsivabalan replied:
yeah. its nothing but the small, medium and large metrics. For eg, incase of getBlobSendTime, it refers to getSmallBlobSendTime, getMediumBlobSendTime and getLargeBlobSendTime. Its not what we were thinking. I was expecting something like sendTime per 100KB or something of those sorts.

pnarayanan replied:
Yeah, which doesn't really make sense for this class though. If you were to describe these variables in this class, how would you do it?
I kinda see how something like `MetricsHistogram` would be useful here. Basically, the caller wants to update two different metrics and that can be encapsulated within `MetricsHistogram` class.
I think what you should do is to keep the things that make sense for `NetworkSendMetrics` within this class. Let `NetworkClient` use this class directly as I don't think it needs any other metric. As far as the server is concerned, it can extend this class to make a server specific `NetworkSendMetrics` class which can override appropriate methods.

nsivabalan replied:
yep, makes sense. I will make changes and will push along with other feedback changes. Guess this change is not very big for you to wait until I update the patch. so, I will hold until I get all reviews (one pass) and will update the patch

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -30,23 +31,30 @@
 public class NetworkMetrics {
```
In general, would prefer grouping these variables by component, rather than by type.

nsivabalan replied:
fine. I will make changes and push in all in one go (along with addressing other feedbacks)

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,65 @@
+   * Updates the time spent by the request or response in the queue before being sent out
+   * @param value the time spent by the request or response in the queue
+   */
+  public void updateQueueTime(long value) {
```
again I think "queue time" does not make sense for `NetworkSendMetrics`. What queueing are we talking about? Would prefer if this class only concerns itself with `NetworkSend` related metrics and the server can extend this class and add other server specific stuff which it can pass on to other components.

nsivabalan replied:
I felt the reasoning as follows. NetworkSend (which encapsulates "Send") is an object which has to be sent across the network. So, irrespective of server or client (or even any other consumer) will queue it up somewhere and then starts to send the "NetworkSend". Selector might take N no of polls(or writes) to complete the actual send. Hence we have 3 metrics, queueTime, sendTime and totalTime.

pnarayanan replied:
It is just a bit weird because `NetworkSend` does not know about any queueing. In fact `NetworkSend` itself never gets queued at least in the `NetworkClient`. It is constructed just before sending.

`NetworkSend` knows about the send becoming complete, because some entity calls `onSendComplete()` on it.

The thing is `NetworkSendMetrics` is not really metrics about `NetworkSend`. A part of it is, and that is the part that `NetworkSend` itself cares about, but the other part is used directly by another entity.

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,65 @@
+   * Updates the time spent by request or response to be completely sent
+   * @param value the time spent by the request or response to be completely sent
+   */
+  public void updateSendTime(long value) {
```
which is why this should be called `onSendComplete()` and is something that the `NetworkSend` will invoke when the send is complete.

nsivabalan commented:
@pnarayanan @xiahome : Can you guys review this.

pnarayanan commented:
will continue with this review as soon as I finish a pass on #326

nsivabalan commented:
@pnarayanan @xiahome : Reminder to review.

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/ClientNetworkRequestMetrics.java:
```diff
@@ -0,0 +1,53 @@
+   * Updates few metrics once the send completes
+   * @param value the time spent by the request to be completely sent out
+   */
+  public void updateSendTime(long value) {
```
`@Override`

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/network/ClientNetworkRequestMetrics.java:
```diff
@@ -0,0 +1,53 @@
+public class ClientNetworkRequestMetrics extends NetworkSendMetrics {
+  private final Histogram requestQueueTime;
+  private final Histogram requestTotalTime;
+  private long timeSpentTillNow;
```
why isn't this final?

pnarayanan replied:
ah, okay, never mind.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -89,8 +96,13 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
       throw new IllegalStateException("The NetworkClient is closed.");
     }
     List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
+    numPendingConnections.set(pendingRequests.size());
```
why set this several times within this method? Just set it once after the for loop.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -89,8 +96,13 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
-      pendingRequests.add(new RequestMetadata(time.milliseconds(), requestInfo));
+      ClientNetworkRequestMetrics clientNetworkRequestMetrics =
+          new ClientNetworkRequestMetrics(networkMetrics.requestQueueTime, networkMetrics.requestSendTime,
+              networkMetrics.requestSendTotalTime, 0);
```
Time spent till now probably could be used in a better way - to track the time spent since the request was created. But we can do that later (I have added it to the backlog)

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -138,17 +153,22 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
             connId = selector.connect(new InetSocketAddress(host, port.getPort()), networkConfig.socketSendBufferBytes,
                 networkConfig.socketReceiveBufferBytes, port.getPortType());
             connectionTracker.startTrackingInitiatedConnection(host, port, connId);
+            requestMetadata.connectionCheckOutAttempts++;
```
this metric does not make sense to me, given how it gets updated.

nsivabalan replied:
which metric are you talking about ?

pnarayanan replied:
the comment is on `connectionCheckOutAttempts`. what is your definition of a connection checkout attempt?

nsivabalan replied:
I have 2 metrics ConnectionCheckOutAttemptsBeforeSucceeding and ConnectionCheckOutAttemptsBeforeFailing to track the no of times we tried to check out for a particular request before we succeed or fail.

pnarayanan replied:
don't you need to update this right after the checkout? why is it being done only when a `connect` is made?

Also, is this metric really useful? You do have a metric for number of connections timing out due to unavailability of connections - is that not good enough? what purpose does knowing the number of attempts made serve - that is totally dependent on the number of times `sendAndPoll()` is called within the network client timeout.

nsivabalan replied:
If you feel, it doesn't add any value, I can remove it

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -138,17 +153,22 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
+              requestMetadata.clientNetworkRequestMetrics, time));
           connectionIdToRequestInFlight.put(connId, requestMetadata);
           iter.remove();
+          numPendingConnections.set(pendingRequests.size());
```
what are we hoping to achieve by updating this metric here? This is not going to be very useful if it gets updated several times within a second. Would suggest updating this once in a `sendAndPoll` - possibly towards the end of it.

pnarayanan replied:
Actually, there is another problem here. `NetworkMetrics` is common to all operation controllers. But there are multiple network clients per operation controller. Based on how we want this, we may have to aggregate these metrics or make them distinct. Right now one overrides the other.

nsivabalan replied:
aah... good catch. Let me think about it. Do you have any inclination towards one of them ?

pnarayanan replied:
Actually I think this is a non-issue. There is no problem with this particular metric actually, as the gauges are per `NetworkClient`. Anything that is not a gauge is common to all, and get aggregated across the `NetworkClient`s, which is probably what you want. But do ensure that.

nsivabalan replied:
Yes, I went over the processor and selector and there are no issues in multiple selectors updating the metrics. Similarly, here even though multiple network clients might update the same metric, no issues, just that we will get an aggregated view

xiahome replied:
@pnarayanan Why
>there are multiple network clients per operation controller
?

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -78,6 +79,26 @@
   // the count of renegotiation after initial handshake done
   public final Counter sslRenegotiationCount;
 
+  // NetworkClient metrics
```
This is not going to work as is as there can be multiple `NetworkClient`s. On that note, how do metrics work for processor/acceptor at the server. All processors share the same `NetworkMetrics`, but initialize individual `Selector`s. Each of those `Selector` call into `initializeSelectorMetrics()` and it looks like only the last selector gets counted towards the metrics.

Could you review that and let us know if there is a problem there? We may have a similar problem here.

pnarayanan replied:
Ah, wait for the gauges there is no problem because a new one is created when those methods are called. So they do not get overridden.

nsivabalan replied:
yes, you are right

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -134,6 +173,19 @@ public Long getValue() {
+   * @param numPendingConnections the count of pending connections to be checked out
+   */
+  void initializeNetworkClientPendingConnections(final AtomicLong numPendingConnections) {
+    networkClientPendingConnections = new Gauge<Long>() {
```
how do these get distinguished though?

nsivabalan replied:
I don't really know. But its on similar lines the selector too updates a gauge.

pnarayanan replied:
We should find out. Should not be hard. Will follow up in person with you.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -160,6 +212,15 @@ public void updateNodeReceiveMetric(String hostName, int port, long bytesReceive
+   * Updates a few metrics when a request is timed-out without successful checkout of a connection
+   * @param connectionCheckOutAttempts number of times, connection check out was attempted before timing out
+   */
+  void updateConnectionTimedOutMetrics(int connectionCheckOutAttempts) {
```
perhaps it would be better to update this directly rather than via this method as this seems like a one-off thing.

nsivabalan replied:
fixed

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -118,6 +131,8 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
             new ResponseInfo(requestMetadata.requestInfo.getRequest(), NetworkClientErrorCode.ConnectionUnavailable,
                 null));
         iter.remove();
+        networkMetrics.updateConnectionTimedOutMetrics(requestMetadata.connectionCheckOutAttempts);
```
basically a method here just to update two metrics seems out of place.

nsivabalan replied:
gotcha

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -159,11 +179,14 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
    *                         the selector events.
    */
   private void handleSelectorEvents(List<ResponseInfo> responseInfoList) {
+    long handleSelectorEventsStartTimeMs = System.currentTimeMillis();
```
What is the use of this metric? There are no I/O or cpu intensive stuff being done within this method.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -107,6 +119,7 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
    * @return the list of {@link NetworkSend} objects to hand over to the Selector.
    */
   private List<NetworkSend> prepareSends(List<ResponseInfo> responseInfoList) {
+    long prepareSendStartTimeInMs = System.currentTimeMillis();
```
again, I think we have to be careful about these metrics to prevent metric overload. We don't really want to instrument every single method for prod code - there should be a reason. These methods do not do any I/O or cpu intensive stuff. For all you know they finish under a millisecond. If we make a case for it here, we could make a case for doing this for every single non-trivial method in ambry.

We could add a metric at a higher level - how long `sendAndPoll` takes, for example, how long operation manager `poll`s take etc. but I don't see a need for having metrics at this level.

nsivabalan replied:
sure. will fix it.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/ClientNetworkRequestMetrics.java:
```diff
@@ -0,0 +1,53 @@
+  private final Histogram requestTotalTime;
+  private long timeSpentTillNow;
+
+  public ClientNetworkRequestMetrics(Histogram requestQueueTime, Histogram requestSendTime, Histogram requestTotalTime,
```
Probably rename it to `ClientNetworkSendMetrics`.
I have a similar suggestion for server, and please see below for more explanation.

nsivabalan replied:
This doesn't capture just send related metrics. It captures all metrics pertaining to a request sent by the client. And hence the name

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,36 @@
+ * Tracks send time metrics for a {@link NetworkSend}
+ */
+public class NetworkSendMetrics {
+  private final Histogram requestOrResponseSendTime;
```
I think you can simply name this metrics `sendTime`.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,36 @@
+public class NetworkSendMetrics {
+  private final Histogram requestOrResponseSendTime;
+
+  public NetworkSendMetrics(Histogram requestOrResponseSendTime) {
```
Same suggestion for renaming `requestOrResponseSendTime`.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  /**
+   * Updates the time spent by request or response to be completely sent
```
Updates sendTime when the networkSend has been sent out completely.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/ServerNetworkResponseMetrics.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Tracks a set of metrics for a network response by a Server
+ */
+public class ServerNetworkResponseMetrics extends NetworkSendMetrics {
```
People should still understand if this is renamed as `ServerNetworkSendMetrics`, which means a `networkSend` sent out by a server. We should not care whether it is a request or a response.

nsivabalan replied:
Again, this is not just capturing metrics for a send. This captures metrics related to a response sent by the server. and hence the name

xiahome replied:
My point is, we only care about this is a `networkSend` sent by a server, but we do not care about if this is a response or a request or something else. So, following this thought, we will have two metric classes, `ClientNetworkSendMetrics` and `ServerNetworkSendMetrics`, which track something sent by a client and something sent by a server. If you see below, the member variables do not depend on whether it is a request or a response.

nsivabalan replied:
@pnarayanan : what is your take on this. We have a split vote, but ok with either ways.

pnarayanan replied:
I think this class is already aware that this is a server response - given how it has a "bySize" metric.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/ServerNetworkResponseMetrics.java:
```diff
@@ -0,0 +1,66 @@
+      Histogram responseTotalTime, Histogram responseSendTimeBySize, Histogram responseTotalTimeBySize,
+      long timeSpentTillNow) {
+    super(responseSendTime);
+    this.responseQueueTime = responseQueueTime;
```
`this.queueTime = queueTime`, same suggestion for the rest of the variables.

nsivabalan replied:
This class deals with responses sent by the server. If we name it just queue time, then it might cause some confusion. So, would like to have it that way

xiahome replied:
I replied with my thought, but I may not fully capture your position. If it is decided the names are kept unchanged, I still suggest rename the variables, since the class name is `ServerNetworkResponseMetrics`, which means everything is about a response. There should no confusion.

xiahome commented on ambry-api/src/main/java/com.github.ambry/network/ServerNetworkResponseMetrics.java:
```diff
@@ -0,0 +1,66 @@
+  @Override
+  public void updateSendTime(long value) {
+    super.updateSendTime(value);
+    if (responseSendTimeBySize != null) {
```
Why this null check is needed? The same question for the code below.

nsivabalan replied:
Some callers might have these metrics and some might pass in null.

xiahome commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -78,6 +79,26 @@
   // the count of renegotiation after initial handshake done
   public final Counter sslRenegotiationCount;
 
```
Line 29. Even until now I do not have a clear idea what a `network layer` refers to. Would be good to specify which part (or what is the focus) for this metrics set.

Line 48. Is `numActiveConnections` common to all selectors? If so, my feeling is that this metric would not be that useful.

Line 35. Is this metrics decremented only in Selector.java line 543? Any exception scenarios would also remove an inflight `send`?

nsivabalan replied:
1. To simply out, anything under network package are considered to be network layer. 
2. @pnarayanan @xiahome : I went over the code once again and realized that there was a bug. Basically the active connections for selector metric was never registered and hence we didn't hit the issue. I have fixing it along with network client's pending connections metric. Do take a look. I am just aggregating all the values from all (selector/networkclient) and update just one metric.
3. Yes, could be. Since this metric is un-related to this patch, can you create a new issue.

pnarayanan replied:
We have an issue for 2: #334. If we are fixing it, please also take a look at whether all the other metrics are getting registered as well, and removing unused methods as described in that issue.

nsivabalan replied:
I will take it up once this patch goes in, I don't wanna drag this anymore. already its been 3 weeks for this patch.

nsivabalan commented:
@pnarayanan @xiahome : Addressed feedbacks. Fixed the gauges for selector active connections and netowrkclient pending connections. 
@pnarayanan : If we want to take the route you suggested(Meter instead of gauge), I would suggest we should revisit all the gauges in a later patch and do a sweeping change. For now, I am following what we have been doing for similar metrics elsewhere.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -85,17 +92,27 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
+        throw new IllegalStateException("The NetworkClient is closed.");
+      }
+      List<ResponseInfo> responseInfoList = new ArrayList<ResponseInfo>();
+      numPendingConnections.set(pendingRequests.size());
```
setting `numPendingConnections` here is unnecessary - it wouldn't have changed at this point. Also why is this called `numPendingConnections`? Is this not `numPendingRequests`?

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -85,17 +92,27 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
+      handleSelectorEvents(responseInfoList);
+      return responseInfoList;
+    } finally {
+      networkMetrics.networkClientSendAndPollTime.update(time.milliseconds() - startTime);
```
I think you should set `numPendingConnections` (which is really `numPendingRequests`) just once - and here in finally. The `NetworkClient` adds all requests to the pending list and then they get a chance to be sent out only when `prepareSends` is called. So these requests are really pending, as far as the metrics are concerned, only if they remain in the pending list even after `prepareSends()` was called.

nsivabalan replied:
I wish to capture at two places. Once at line 107 after adding new requests. And once in finally. I will remove all other instances.

pnarayanan replied:
You shouldn't at line 107, and that is my point. In order for the metric to make sense, these requests should not be considered as pending unless they remain in the pending list after the prepare.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -140,15 +158,18 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, int maxConn
         logger.error("Received exception while checking out a connection", e);
       }
     }
+    numPendingConnections.set(pendingRequests.size());
```
so this is not necessary either - this metric should simply be updated at the end of `sendAndPoll`

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -163,7 +184,9 @@ private void handleSelectorEvents(List<ResponseInfo> responseInfoList) {
 
-    for (String connId : selector.disconnected()) {
+    List<String> disconnectedList = selector.disconnected();
+    networkMetrics.connectionsDisconnectedCount.mark(disconnectedList.size());
```
this seems like it should be part of the selector metric - it there one already for this?

nsivabalan replied:
we have one already selectorDisconnectedErrorCount in NetworkMetrics. Should I remove this one ?

pnarayanan replied:
Yes, I think this is redundant.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -191,17 +215,39 @@ public void close() {
     // the RequestInfo associated with the request.
     RequestInfo requestInfo;
+    // the time at which this request was queued.
+    private long requestQueuedTimeMs;
```
I know this is old code, but it was a bit confusing as this seemed like elapsed time. Do you think you could call this `requeustQueuedAtMs` and `requestDequeuedAtMs`?

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -191,17 +215,39 @@ public void close() {
+     * Actions to be done on receiving response for the request sent
+     */
+    void onResponseReceive() {
+      networkMetrics.requestResponseRoundTripTime.update(System.currentTimeMillis() - requestSentTimeMs);
```
The router tracks `requestLatency` which should be a function of this and `requestQueuedTime` I would think. Just noting, it is useful to track this.

nsivabalan replied:
Yep. So, we might have to modify ResponseInfo to include these latency values right? do we have a ticket to work on this ?

pnarayanan replied:
That was just a note. This is already being done as part of #326.

nsivabalan replied:
cool.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -118,20 +133,52 @@ public NetworkMetrics(MetricRegistry registry) {
+   * @param numActiveConnections count of current active connections
+   */
+  void registerSelectorActiveConnections(final AtomicLong numActiveConnections) {
+    Gauge<Long> selectorActiveConnectionsCount = new Gauge<Long>() {
```
I think your original idea of grouping this together is better - there is not much insight in having selector level granularity for this metric. Will follow up with you.

pnarayanan commented on ambry-network/src/main/java/com.github.ambry.network/NetworkMetrics.java:
```diff
@@ -118,20 +133,52 @@ public NetworkMetrics(MetricRegistry registry) {
-  void initializeSelectorMetrics(final AtomicLong activeConnections) {
-    numActiveConnections = new Gauge<Long>() {
+  void registerNetworkClientPendingConnections(final AtomicLong numPendingConnections) {
+    Gauge<Long> networkClientPendingConnectionsCount = new Gauge<Long>() {
```
same here

pnarayanan commented:
Looks good overall. Can merge once you address comments. Please also rebase with master (this branch has conflicts).

xiahome commented:
LGTM except my two pending comments. They are anyway naming suggestions so I am not pushing for a change if both @nsivabalan and @pnarayanan feel ok.

nsivabalan commented:
@pnarayanan @xiahome : rebased and addressed the comments

pnarayanan commented:
LGTM. Could you confirm that this builds on Linux and is formatted correctly?

nsivabalan commented:
yes. confirmed.
